Ffmpeg hwaccel

mcosmin222 opened the pull request:
This PR adds HW acceleration support for ffmpeg software decoder (or tries to anyway).

Progress so far:
Video playback basically works.

Issues:

1. The seek bar doesn't seem to move for some reason.
2. Some frames still report 0 duration, even for video.
3. It only supports hw methods which require "hw device context", not "hw frame context" -> I basically just hope all of them are like that. 
4. I essentially removed software decoding completely, we will need to create different paths for software and hardware (different get_format methods maybe?)
5. Generally, the code will need improvements, maybe try to avoid the whole "copy from GPU to CPU part", which is quite a headache
6. Still haven't figured out how to use the proper hardware device pointer. My guess (and hope) is that we can use the proper hardware device pointer to create the surfaces directly over the frame data.
7. I only modified the x64 build to support hwaccel. I'm not even sure if this makes sense for 32 bit applications though, some hardware decoders are 64 bit only AFAIK.

So lots more to do here...

lukasf commented:
Cool, good work man!

I read a bit more into docs. In the description of "av_hwdevice_ctx_create()" it says:

"This is a convenience function intended to cover the simple cases. Callers who need to fine-tune device creation/management should open the device manually and then wrap it in an AVHWDeviceContext using av_hwdevice_ctx_alloc()/av_hwdevice_ctx_init()."

So I think it should hopefully not be be too difficult to get our device into the toolchain. But it's probaby not the top priority.

I will merge #173 soon, then I will start playing with this HW branch.

mcosmin222 commented:
You know the funny part is that I didn't check for store compliance on the dlls...

lukasf commented:
I merged with master and managed to build and run this: Playback works on GPU, but I get green tint all over the image. Probably something wrong with the PixelType. Also seeking does not work and 0 duration on all frames. We need to check where that comes from. Maybe we need to copy over the duration from some other place (original hw frame before copy, or AVPacket).

I ran AppCertKit and there are no API violations. So I think we can proceed with this.

Will cleanup a bit and push master branch version, hopefully tomorrow.

mcosmin222 commented:
I did the checked the compliance too, indeed no violation.

Interesting about the green tint. I didn't get that myself, i guess it depends on the file being used.

lukasf commented:
So I have updated the branch to master. I built a HW acceleration enabled (all platforms) prerelease NuGet package, it is already referenced. Cool thing: Now anyone can test the HW branch without building FFmpeg! I will try to come up with some docs on how to create and reference local NuGet packages, to make it easy for devs to create and use their own FFmpeg builds, until an official one is available on NuGet source.

I also fixed the duration issue, which fixed the seek issue as well.

I get the green tint on almost all files. They all use NV12 pixel format. I have only one file here that does not have green tint, it is a 4K UHD file with P010 format. It plays correctly, but it is horribly slow!! I noticed that we actually copy the data twice: Once from HW frame to SW frame, and after that, the data is copied again by SWS.

Our lib already supports 10bit output through P010 format. So I think it should be possible to get rid of the second copy by SWS. But maybe we should instead go directly for pure GPU samples? Let's see what we can do here.

mcosmin222 commented:
Cool. Nice work!

Ironically, I did test with a 4k UHD file ^^

I was looking at this
https://docs.microsoft.com/en-us/uwp/api/windows.graphics.directx.direct3d11.idirect3dsurface?view=winrt-19041

So looking at 
https://docs.microsoft.com/en-us/windows/win32/api/d3d11/nf-d3d11-id3d11device-createtexture2d

We can create a texture2D from existing data.

The existing subresource can be initialized like this

https://docs.microsoft.com/en-us/windows/win32/api/d3d11/ns-d3d11-d3d11_subresource_data

However, the parameter is suspiciously called "pSysMem" which points to this being CPU data.

mcosmin222 commented:
Interesting
The hardware av frame data pointers can be converted directly to IDirect3DSurface 

auto resource = reinterpret_cast<ID3D11Texture2D*>(avhwFrame->data[0]);

seems to succeed, and I can then query the Description in there.

![image](https://user-images.githubusercontent.com/9284164/89550200-000bb680-d812-11ea-9acc-f120caa614a4.png)

lukasf commented:
Yes this is what I read as well. The texture can be cast to IDXGISurface and then we can create the IDirect3DSurface from it using CreateDirect3DSurface(). But we need to create our own HW device context first and fill with our app's device.

mcosmin222 commented:
Ok then, I will research that instead ^^

mcosmin222 commented:
so annoying, when I include 
#include <windows.graphics.directx.direct3d11.interop.h>


I get compiler errors
error C2872: 'EventRegistrationToken': ambiguous symbol
and some others

mcosmin222 commented:
So I just tried it out for the sake of it, passing down the surfaces I get from the av hw frame and I get a green image.
When you tried it the first time, did you get the same green image, or no image at all?

Maybe we do not need to actually get that app device pointer.

mcosmin222 commented:
I am fairly certain now that the only way to do this is to create a new texture with the our device handle and copy / wrap the hw texture of the Av hw frame.

lukasf commented:
I did not try this, just did some investigations / docs reading.

mcosmin222 commented:
I pushed a few new changes, this is my attempt to copy the hw frame from one device to another, but for some reason, the DXGI surface on our frame is NULL, thus we cannot do the surface to media sample ....

I don't see where i'm doing wrong.

mcosmin222 commented:
Well, I got past that point, i was using the wrong texture description.

Still - a green screen. 
The only strange thing is that the description of the ffmpeg texture has array size of 32, but when I try to create a texture with array size of 32 to copy to, it will fail, it forces the array size to be 1 in order to succeed.

mcosmin222 commented:
I have succeeded.
No green tint I think...

mcosmin222 commented:
So I think the next step is to properly switch between hardware and software decoding depending on hardware capability. We might not even need pass through anymore,since I think we now have the complete pipeline on our side of the code.

lukasf commented:
Nice work.

I have been trying to get decode on the actual MSS GPU to work, without texture copy. But it looks like this is not fully possible without changes in FFmpeg itself. Generally, we can create and assign the HW device context later, after creating and opening the AVCodecContext. It is only used when the first frame is actually decoded. So I manually created and initialized a HW device context in OnStarting, using the device pointer from our MSS. That part worked. But the problem is this: FFmpeg currently creates an array texture for its buffer pool. But an IDXGISurface is only created for textures with only a single item. So when I get the ID3D11Texture2D and QueryInterface for IDXGISurface, it fails with E_NOINTERFACE. There does not seem to be a way to work around this. I digged around and found a method CreateSubresourceSurface which accepts an IDXGIResource (I can get this from the texture) and an index. This successfully returns an IDXGISurface from the array ID3D11Texture2D. But creating the MSS sample with this surface fails, also with E_NOINTERFACE - probably because this IDXGISurface does not implement ID3D11Texture3D (I tried QueryInterface - no interface). So as long as FFmpeg only works with array style textures, we won't be able to get rid of the texture copy step.

Theoretically, it would be possible to change FFmpeg to work with normal non-array textures. A few parts are kind of prepared for this, it wouldn't be such a big change. But I don't like the idea of starting to patch around in FFmpeg. And getting this into official FFmpeg is probably a long way to go. So I don't see this happening, unless someone has a good idea how to do this.

Maybe we could still use the approach of late-assigning the HW device context. Then we are sure that the decode happens on the actual app GPU (there could be a situation where the app does not run on the default device). We'd still have to copy the texture, but it could be simplified a bit (no shared texture stuff required).

mcosmin222 commented:
Oh so you dug pretty deep. Nice.

I don't think we should patch ffmpeg itself. It will be difficult to maintain it in case others do changes there, and according to hwaccel page, they are still work on some stuff. 

Hace you tried the directx interop helper I pushed? That converts IDXGISurface to Idirect3DSurface.

lukasf commented:
The problem is that I can't get a valid IDXGISurface. The texture does not return it on QueryInterface, because it will only be returned for textures with ArraySize=1 (this is also mentioned in the docs). There does not seem to be a way to get a valid IDXGISurface for array textures. I could get one through CreateSubresourceSurface, and I can create a IDirect3DSurface from that, but it won't implement ID3D11Texture2D, and creating a sample from that IDirect3DSurface will fail with E_NOINTERFACE. Seems like MediaStreamSample expects the passed surface to also implement ID3D11Texture2D, otherwise it will fail.

My code is pretty messy right now. Will try to clean up and push the changes, but it could take a day or two.

mcosmin222 commented:
I think the reason ffmpeg used the array texture is because Microsoft essentially tells you to do so

https://docs.microsoft.com/en-us/windows/win32/medfound/supporting-direct3d-11-video-decoding-in-media-foundation#allocating-uncompressed-buffers

This seems to be requirement for the decoders, to have several arrays to using for buffering and deinterlancing.

> 
> Create a 2D texture array by calling ID3D11Device::CreateTexture2D. In the D3D11_TEXTURE2D_DESC structure, set ArraySize equal to the number of surfaces that the decoder needs. This includes:
> 
>     Surfaces for reference frames.
>     Surfaces for deinterlacing (three surfaces).
>     Surfaces that the decoder needs for buffering.
> 
> The binding flags (BindFlags) should include the D3D11_BIND_DECODER flag and any bind flags set through the MF_SA_D3D11_BINDFLAGS attribute in the output stream attributes.

lukasf commented:
Yes I have read that part as well. It's probably most efficient to allocate all textures in one go. But that does not neccessarily mean that it wouldn't work when allocating textures separately. Since there is a code path for allocating singular D3D11 textures in FFmpeg (although not used), my guess is that this would actually work.

The annoying truth is: The Win32 MF APIs for creating samples have full support for array style textures, passing the texture and an index. Only the UWP wrappers do not support this.

mcosmin222 commented:
I'm still quite surprised this works at all with UWP...
copying the frame seems to be just a minor annoyance at this point.

mcosmin222 commented:
Meh I broke the frame grabber, probably the subtitle provider as well...

mcosmin222 commented:
I wonder if ffmpeg can automatically switch between HW and SW decoding depending on GPU capabilities, by using the get_format method for determining pixel format. This could save us a lot of time when doing the switching.

mcosmin222 commented:
I have also tried to get the MSS GPU pointer to work with ffmpeg, much the way you described. Initialization seems to work fine, but when I try to cast the AvFrame data pointers to texture 2D, it returns a bogus pointer, I can't get the texture description or use query interface on it.

Was there anything else that you've done aside from assigning the device, video service, and their contexts to the av hw device context?

mcosmin222 commented:
> I wonder if ffmpeg can automatically switch between HW and SW decoding depending on GPU capabilities, by using the get_format method for determining pixel format. This could save us a lot of time when doing the switching.

It appears so.

mcosmin222 commented:
Hi @lukasf,
Sorry about my rumblings here ^^

I did manage to get it working in the MSS GPU pointer after all. This seems to have improved performance by a bit ^^.

lukasf commented:
I fixed up SW processing, so now it works with both SW and HW decoding.

Did you notice that you can only open one file? When the second file is opened, access violations occur, even when we don't use HW decoding. I tried to work on this, but without success.

Second problem I noticed: When the app is off screen for a few seconds, errors occur because the device is halted. We cannot create new textures in that state, the texture created is NULL. I guess this could be solved if we create a texture pool (and fill upfront) for our copy textures. That would be better for the GPU as well, I guess.

mcosmin222 commented:
I will check on these.

One question though : why do we not support VLD?

lukasf commented:
VLD is an onld, deprecated decoder in FFmpeg (and it does not work). It is recommended to only use D3D11.

mcosmin222 commented:
> Did you notice that you can only open one file? When the second file is opened, access violations occur, even when we don't use HW decoding. I tried to work on this, but without success.
> 
> Second problem I noticed: When the app is off screen for a few seconds, errors occur because the device is halted. We cannot create new textures in that state, the texture created is NULL. I guess this could be solved if we create a texture pool (and fill upfront) for our copy textures. That would be better for the GPU as well, I guess.

I think the problems are related.
It also seems to happen when seeking too far from the current position, although this one is a bit random.

One thing I discovered though: when opening the 2nd file, it actually does get to play for a bit, and then gives off the crash, this might be some winRT lifecycle shenanigans (aka bug).

mcosmin222 commented:
Interesting.
I checked-out the commit 217b4a3 which is right before using the same device pointer in mss for ffmpeg, and I can get neither of the 2 issues. 

It seems the problem is in ffmpeg, or something that changed between that commit and latest version.

Also memory consumption has a pretty weird pattern

![image](https://user-images.githubusercontent.com/9284164/90607788-6200ee80-e20a-11ea-9f76-818cecf8acc8.png)

Seems ffmpeg will buffer textures internally, and release them when the memory is full.

Can you confirm the same @lukasf ?

lukasf commented:
FFmpeg uses a fixed size texture array. The memory usage most probably comes due to us allocating a new texture for every sample. Somehow they are also not properly released as it seems. Memory usage would sure be a whole lot better if we'd use a texture pool, recycling textures on SampleProcessed event.

About opening a second file, there could be a problem with FFmpeg automatically releasing our device pointer when MF/UWP still holds some samples. Releasing the samples later could cause trouble. That's just me guessing, of course.

mcosmin222 commented:
So in the end, it seems to be more stable if we use 2 device pointers. Should we go forward with this?

lukasf commented:
In that older solution, we are getting the device and device context from MSS, but both are never released, leaking memory. Once we'd properly release them in the destructor, the same issue might occur.

Also that solution would most probably have a high performance impact, when there are two GPUs and the MSS runs on a different GPU than FFmpeg. As long as it is only two different pointers to the same device, it's probably not that much of a difference, but there could be other situations.

mcosmin222 commented:
@lukasf 

This last commit seems to have fixed the stability problem for me. Can you confirm?

lukasf commented:
Are you sure you disabled HW acceleration in the sample tab? For me, it still crashes after opening the second file.

mcosmin222 commented:
Yes. Are you trying this with or without a debugger attached?

lukasf commented:
It seems that it's not possible to automatically keep references alive. I tried multiple approaches and it all failed. But it was pretty easy to undo my last fix, and instead just fix up the samples. We just have to make sure that our instance is referenced until MediaOpened of the next file. In the older CPP sample, we'd lose the reference to our old FFmpegInteropMSS right after assigning the next source, but at that point, the old MediaStreamSource is still playing. That was a pretty dangerous thing anyways. Now I keep the old reference alive as actualFFmpegMSS, until the next is completely opened. I did the same in the C# sample, since under high GC pressure, the same crash might occur.

mcosmin222 commented:
Feels pretty stable now.

lukasf commented:
I am surprised how stable it is. Didn't we get this crash when the app is off screen for a while? I don't remember doing anything about that. Can you check if it still happens on your machine?

I still think that it would be better to use a texture pool, instead of allocating new textures all the time.

villagra commented:
I'll try with a bunch of files once i return from holidays... can't do anything from here. awesome work guys

mcosmin222 commented:
@lukasf 

No, it doesn't crash anymore in background. Maybe it had to do with memory pressure since we weren't releasing samples correctly.

lukasf commented:
I added a texture pool. Seems to work fine so far. 5 textures are needed for normal rendering.

One new issue I noticed: The new approach will cause high GPU memory usage when MSS buffer is used, because then, decoded samples are buffered, all in GPU memory. So in streaming scenarios, when a large buffer is needed, we could run into problems. Here the old passthrough approach is more memory efficient, because it buffers small encoded packets instead of large decoded samples. Sooner or later we really need to think about adding our own internal read-ahead buffering on AVPacket level. Until that is available, we have to keep the old passthrough approach as alternative decoding path.

mcosmin222 commented:
Nice work!

But I am wondering : how did you figure 5 are enough?

lukasf commented:
I read in the docs for DXVA that the EVR renderer (also used by UWP) requires 5 textures for its buffering. Also FFmpeg adds 5 textures for rendering to its array. And my testing confirmed that exactly 5 textures are requested, before the first is returned. Only when I add MSS BufferTime, more textures are requested. In that case, my pool will dynamically allocate more textures.

10 seconds buffer almost eats 1GB of GPU memory!

mcosmin222 commented:
Looks good.

About the buffer time: I think it is sufficient for now to just keep the passthrough in, although this probably bypasses the paid extensions from the MS store.

Do you think it would be difficult to implement the buffering on "compressed" side of the decoder?

Unfortunately, I won't have much time for development in the coming days... should we merge this and implement the buffering at a later date?

lukasf commented:
Implementing an async read-ahead buffer is a non-trivial task. We should definitely do that as a separate PR, when we have time for it. It is only really needed for streaming scenarios, and we still have our passthrough as workaround available.

Currently, I have enabled the new HW acceleration by default. We should be aware that this could be a breaking change for streaming scenarios with very high buffer times.

The enum values for VideoDecoderMode are a bit confusing right now. My proposal is this:
- Automatic: This is the new default
- AutoDetection: Gets deprecated because from the name you don't know the difference to Automatic
- AutomaticSystem: Would be the new replacement for AutoDetection
- ManualSelection: I would like to deprecate the whole method of manual passthrough selection. So this enum value would get deprecated and all corresponding video passthrough properties in the config class are deprecated. New preferred way would be using either Automatic, AutomaticSystem, or one of the Force... values
- ForceSystemDecoder: No change
- ForceFFmpegSoftwareDecoder: No change

There is also one TODO in the code: There is currently no way of knowing upfront, if FFmpeg will really select HW acceleration for a stream (e.g. wrong encoding profile or too high resolution). In case HW acceleration is not used (Line 82 in D3D11VideoSampleProvider), we need to propagate this to the StreamInfo. We only really know if HW acceleration is used, after the first sample of a stream has been decoded.

mcosmin222 commented:
I agree and the enum values. I personally got confused as well, i checked the code paths to see which goes where. The proposal looks fine, maybe AutomaticSystem should be "AutomaticPassthrough"?

mcosmin222 commented:
Ah you already did that.

lukasf commented:
The info in HW decoder availability is only general for the codec. It will not respect resolution, encoding profile or number of b-frames of the actual file. That is only later checked during open. No HW format will be offered in get_format function if the encoding is not compatible, so in that case we will "fallback" to SW decoding.

lukasf commented:
For the term "passthrough", I'd like to avoid that in newer APIs. Passthrough in media playback is usually associated with untouched passthrough of DTS or AC3 streams to an external AV receiver or TV set. That's why I switched to using "System" or "SystemDecoder", as opposed to FFmpegSoftwareDecoder or HW decoder.

mcosmin222 commented:
Alright, then it is settled. 
I suppose the old enum will stay until at get to remove pass through all together, but issue a warning, right?

lukasf commented:
Yes, when we deprecate the enum value, a warning will be shown on use.

lukasf commented on Samples/MediaPlayerCPP/MainPage.xaml:
```diff
@@ -49,7 +49,7 @@
 
                         <ToggleSwitch Header="Auto-Detect Hardware Accelection" OffContent="OFF" OnContent="ON" x:Name="AutoDetect" Toggled="AutoDetect_Toggled" />
-                        <ToggleSwitch Header="Passthrough Video" OffContent="OFF" OnContent="ON" IsOn="True" x:Name="PassthroughVideo" Toggled="PassthroughVideo_Toggled" Margin="0,0,0,20" />
+                        <ToggleSwitch Header="Passthrough Video" OffContent="OFF" OnContent="ON" x:Name="PassthroughVideo" Toggled="PassthroughVideo_Toggled" Margin="0,0,0,20" />
```
Might want to rename to "Force system video decoder"

lukasf reviewed (changes requested):
I found a few minor issues during code review, will solve them now. @mcosmin222 can you also do a thorough review? It's a pretty big changeset.

mcosmin222 commented:
I will take a look on Friday.

mcosmin222 commented on FFmpegInterop/FFmpegInteropMSS.h:
```diff
@@ -345,6 +366,12 @@ namespace FFmpegInterop
 		bool isFirstSeek;
+		AVBufferRef* avHardwareContext;
+
+		bool IsAVSampleSource()
```
I used this to bypass hw accel when extracting frames.
We do need to use it since now the frame grabber is broken.

lukasf replied:
I already fixed FrameGrabber in a different way.

mcosmin222 commented on FFmpegInterop/D3D11VideoSampleProvider.h:
```diff
@@ -0,0 +1,202 @@
+		TexturePool^ texturePool;
+		std::vector<MediaStreamSample^> samples;
+
+};
```
Seems formatting is off here.

mcosmin222 commented on FFmpegInterop/FFmpegInteropMSS.cpp:
```diff
@@ -1192,12 +1209,48 @@ MediaSampleProvider^ FFmpegInteropMSS::CreateVideoStream(AVStream* avStream, int
+				{
+					if (config->pix_fmt == AV_PIX_FMT_D3D11 && config->methods & AV_CODEC_HW_CONFIG_METHOD_HW_DEVICE_CTX)
+					{
+						if (!avHardwareContext)
```
I guess we will not be supporting hw frames method?

mcosmin222 commented on FFmpegInterop/FFmpegInteropMSS.cpp:
```diff
@@ -1515,6 +1578,39 @@ void FFmpegInteropMSS::OnStarting(MediaStreamSource^ sender, MediaStreamSourceSt
+	{
+		ID3D11Device* device = nullptr;
+		ID3D11DeviceContext* deviceContext = nullptr;
+
```
Shouldn't we release those pointers?

lukasf replied:
Frankly, the pointer handling is not very well done right now. We put both pointers into the hw device context. When the hw device context is released, the pointers are released. The better way would probably be to AddRef and Release them wherever we store them, that is how COM pointers are supposed to be used. Or maybe even better, we should be using ComPtr<>, which would automatically handle AddRef and Release for us.

lukasf replied:
I tried with ComPtr, but it turned out tricky and the docs are very poor. In a lot of situations I was not sure how exactly to work with them. So I reverted and use classic AddRef/Release. When working with these COM interfaces, anyone taking ownership (store in a field) needs to AddRef and later Release. Now that I do it correctly in all placed, I don't see anymore crashes.

mcosmin222 commented on FFmpegInterop/FFmpegInteropMSS.cpp:
```diff
@@ -1549,7 +1645,8 @@ void FFmpegInteropMSS::OnSampleRequested(Windows::Media::Core::MediaStreamSource
 		else if (currentVideoStream && args->Request->StreamDescriptor == currentVideoStream->StreamDescriptor)
 		{
-			args->Request->Sample = currentVideoStream->GetNextSample();
+			auto sample = currentVideoStream->GetNextSample();
```
This probably needs to be reverted, I just separated it to be able to put a breakpoint and detect null samples easier.

mcosmin222 commented on FFmpegInterop/TexturePool.h:
```diff
@@ -0,0 +1,133 @@
+		}
+
+	private:
+		ID3D11Device* device;
```
Shouldn't this be release in destructor?

lukasf replied:
I think it works, but pointer handling could sure be done better.

lukasf replied:
Now it is properly released.

mcosmin222 commented on FFmpegInterop/UncompressedSampleProvider.cpp:
```diff
@@ -18,9 +18,13 @@
 
 #include "pch.h"
 #include "UncompressedSampleProvider.h"
+#include <EventToken.h>
```
Unneeded include directive?

mcosmin222 commented on FFmpegInterop/UncompressedVideoSampleProvider.cpp:
```diff
@@ -151,7 +152,7 @@ IMediaStreamDescriptor^ UncompressedVideoSampleProvider::CreateStreamDescriptor(
 
-HRESULT UncompressedVideoSampleProvider::InitializeScalerIfRequired()
+HRESULT UncompressedVideoSampleProvider::InitializeScalerIfRequired(AVFrame* avFrame)
 {
```
I made this change when we copied the hw frame to sw frame, in order to ensure that whatever the hw pixel format was, the sw frame was always the right output. I do not believe this is needed anymore, and can be reverted.

mcosmin222 commented on Samples/MediaPlayerCS/MainPage.xaml.cs:
```diff
@@ -328,13 +329,17 @@ private void Track_Resolved(TimedTextSource sender, TimedTextSourceResolveResult
         }
 
         private async void DisplayErrorMessage(string message)
         {
```
Should return async task

mcosmin222 reviewed (commented):
Right, so the big take away is that the frame grabber no longer works ^^

mcosmin222 commented:
Unfortunately I found a video which misbehaves on hw accel

![image](https://user-images.githubusercontent.com/9284164/92225078-66d0ce00-eeab-11ea-8856-177568091397.png)

I think key frames are fine, but the image degrades over time.
I will try to upload it...

mcosmin222 commented:
@lukasf 

Please provide a (disposable) email address so I can send you the download link.

thanks.

lukasf commented:
send to: tempvid@you-spam.com

mcosmin222 commented:
Sent.

lukasf commented:
On my PC, that file plays fine. Maybe your GPU does not like it. Can you check with a different player? E.g. MPC-BE will show a "GPU" symbol when HW acceleration is used, so you know if it's really HW accelerated or not. First I thought it has too many ref frames, but number of ref frames is ok because it has black bars cropped.

Unfortunately, I still get occasional crashes when I repeatedly open files very fast (in the sample app, press Ctr+Return every second). I've seen this since some time, still happening with latest changes. Usually the stack trace shows FFmpeg calling into d3d11 when we send a packet to the decoder, which results in an access violation. But sometimes there is no d3d11. I'm not sure if it will happen on normal use, still this is a bit disturbing. It is easier to reproduce in the CPP sample, but I've also seen crashes today in the C# sample when excessively opening files.

mcosmin222 commented:
@lukasf, I forgot to mention this happens after a seek.
MPC BE is fine.
Also the first few minutes of that file aren't even rendered. Are you sure you tried the correct one? The "star wars" is causing the issue.

mcosmin222 commented:
We should pick another shortcut for that reload video thingy, cause other elements steal the focus every time.

lukasf commented:
Weird. I also play on Intel HD 630:

![grafik](https://user-images.githubusercontent.com/2306201/92328214-4fd2dd00-f05f-11ea-8c4f-cc0d46138b04.png)

All fine, also after seek. Same on my GTX 1050.

At least, I am glad to tell you that I fixed the crash. It was caused by incorrect pointer handling. MediaSampleProvider was calling Release although it never called AddRef. Now it is super stable!!

lukasf commented:
I have to add that my download got interrupted after half of the file. It still plays and seeks (first half of it). But maybe things would be different if I had the complete file. Thing is, I am currently on a rather slow (LTE based) internet connection, so large downloads are a bit tricky for me.

Can you try to create a sample using ffmpeg like this:

ffmpeg -i 'Star.Wars.Episode.VI.Return.of.the.Jedi.1983.720p.BluRay.DTS-ES.x264-DON.mkv' -t 00:03:00 -c copy 'Star.Wars.Test.mkv'

This will create a sample of the 3 first minutes. If the sample shows the same corruption for you, please upload into the same folder.

mcosmin222 commented:
> At least, I am glad to tell you that I fixed the crash. It was caused by incorrect pointer handling. MediaSampleProvider was calling Release although it never called AddRef. Now it is super stable!!

Nice @lukasf ! In all fairness, I think it was my fault, I added AddRef/Release at some point, and forgot to remove the "Release".

I did the trimming as you asked, and the issue is still there. I have uploaded it to the folder.

lukasf commented:
Also with your sample, playback is perfectly fine. First 4 seconds are black, then the 20th century fox animation comes, then lucasfilm logo, then the movie starts. No artifacts during normal playback or after seek. It's pretty weird that we get such different behavior on same iGPU type. Even driver issue is unlikely, since you get the same issue on the NVidia GPU but on my side it plays fine on NVidia as well. Maybe OS version? I am on Windows 19041.450.

mcosmin222 commented:
Hi @lukasf 

I have the same windows version, just updated drivers, still happens.
And it only happens with our ffmpeg hwaccel implementation, software and passthrough works fine.

But if this were to happen on your PC, what would you do to fix this?

mcosmin222 commented:
I get some errors in the output window

FFmpeg (Error): [AVHWFramesContext @ 0000024BE3FE20C0] Static surface pool size exceeded.

FFmpeg (Error): [h264 @ 0000024BE4495400] get_buffer() failed

FFmpeg (Error): [h264 @ 0000024BE4495400] thread_get_buffer() failed

FFmpeg (Error): [h264 @ 0000024BE4495400] decode_slice_header error

FFmpeg (Error): [h264 @ 0000024BE4495400] no frame!

....................

FFmpeg (Error): [AVHWFramesContext @ 0000024BE3FE20C0] Static surface pool size exceeded.

FFmpeg (Error): [h264 @ 0000024BE4498580] get_buffer() failed

FFmpeg (Error): [h264 @ 0000024BE4498580] thread_get_buffer() failed

FFmpeg (Error): [h264 @ 0000024BE4498580] decode_slice_header error

FFmpeg (Error): [h264 @ 0000024BE4498580] no frame!

 - QueuePacket
 - PopPacket
FFmpeg (Error): [AVHWFramesContext @ 0000024BE3FE20C0] Static surface pool size exceeded.

FFmpeg (Error): [h264 @ 0000024BE4495880] get_buffer() failed

FFmpeg (Error): [h264 @ 0000024BE4495880] thread_get_buffer() failed

FFmpeg (Error): [h264 @ 0000024BE4495880] decode_slice_header error

FFmpeg (Error): [h264 @ 0000024BE4495880] no frame!

.....................

FFmpeg (Error): [AVHWFramesContext @ 0000024BE4912500] Static surface pool size exceeded.

FFmpeg (Error): [h264 @ 0000024BE4495D00] get_buffer() failed

FFmpeg (Error): [h264 @ 0000024BE4495D00] thread_get_buffer() failed

FFmpeg (Error): [h264 @ 0000024BE4495D00] decode_slice_header error

FFmpeg (Error): [h264 @ 0000024BE4495D00] no frame!

Decoder failed on the sample.

mcosmin222 commented:
Interesting.
What CPU do you have?

With multithreading enabled

![image](https://user-images.githubusercontent.com/9284164/92384176-40639a80-f118-11ea-9094-28a6bdf1e7f9.png)

With multithreading disabled

![image](https://user-images.githubusercontent.com/9284164/92384223-5cffd280-f118-11ea-882e-221e5f0a676c.png)


This puny line destroys everything for me ^^. If I comment that, it works correctly. 
s->thread_count = 1;

mcosmin222 commented:
It seems the problem is when we set that value for disabling multithreading for hw accel. If I set the thread count to 1 from the start, using the config->MaxVideoThreads, and gets set in

> 	avVideoCodecCtx->thread_count = config->MaxVideoThreads == 0 ? threads : min(threads, config->MaxVideoThreads);
> 				avVideoCodecCtx->thread_type = config->IsFrameGrabber ? FF_THREAD_SLICE : FF_THREAD_FRAME | FF_THREAD_SLICE;

then it also works correctly. But when it was set in the get_format method, it bugged things out, because it was called after the above code was also called.
 
Either you are using a different version of ffmpeg, or there is a hardware difference thing here that caused the issues on my machine.

Is there a particular reason why we are disabling multithreading for HW decoding? DirectX 11 should be pretty multithreading safe.

lukasf commented:
Wow, nice catch! I never would have assumed that this caused it.

I disabled multithreading for the following reasons: First, it does not make sense to use multithreading when HW acceleration is used. Decoding happens on the GPU, not on the CPU, and the GPU will decode on its HW decoder unit and will parallelize internally. Second: When I tested this and debugged into FFmpeg, I noticed that the HW decoder will allocate a separate frames context for every thread, although the frames context is supposed to be shared afaik. When a frames context is created, it does already respect the number of threads, and allocates one additional texture per thread. But creation of the frames context does not happen once, it happens n times (per thread). So FFmpeg allocates way too many textures, unneccessarily eating GPU memory.

I read some entries on the FFmpeg mailing list / bug tracker on this issue. It turns out that the decoder is supposed to internally disable multithreading. This was implemented at some point, according to discussion. But it does not seem to work. So multithreading is really used when threads > 1, although it does not make any sense, and then additionally, the implementation is buggy, at least, it eats way more GPU memory than required. The recommendation is to generally disable multithreading when HW decoding is used. But we only really know if HW decoding is used, when we actually start decoding a frame!? So I tried to set threads to 1 in get_format, which gets called during first decode. It worked fine for me. It's interesting that you get problems with that.

How many logical cores does your device have? I have 8 logical cores (4 physical).

mcosmin222 commented:
12 logical cores, 6 physical cores.
Don't we already know what formats are supported by hardware decoding? Can't we use codec checker to know if hardware will be used to decode?

lukasf commented:
Huh. Maybe setting thread_cout to 1 in that place does not even work. I am not 100% sure if I verified it.

We know which formats are generally supported. But FFmpeg will check at least encoding profile and probably also level and number of ref frames, plus resolution. FFmpeg will detect unsupported files and not offer HW acceleration in get_format in that case (I think I already saw this with one file). I am not sure if it is a good idea to try and mirror the FFmpeg logic on our side. Also, we'd better use the same device instance for our decision, if we'd follow that approach (current CodecChecker will use the default device, which might be different than the MSS one).

Perhaps there could be a different approach (maybe in addition to clever guessing): When we assume that HW acceleration is assumed, we set thread_count to 1 before opening the codec. When the first frame is decoded and HW acceleration was not selected, then we could close the codec and open a new one, this time with multithreading enabled. This would get us back to full multithreading. Not sure if that is a good idea, though. Maybe we have to give this some more thought.

mcosmin222 commented:
How about we do the reverse: start with 1 thread, and set the max number when software is chosen?

I did some preliminary tests with this and it seems to work. I can see why upping the number of threads is safe, while lowering it wouldn't be.

I will need to test it some more though....

lukasf commented:
Are you sure that multithreading is really used when you are upping threads there? Also, the problem is: We cannot access our config class from get_format.

lukasf commented:
Oh man, why the fuck can't Microsoft stop breaking builds on VS updates?! I wanted to rebuild FFmpeg to allow me to debug. Only to find out that build is broken. MS suddenly does not like the vcxproj files of our SMP libs, that were previously working fine. I fucking hate this. I only updated VS because I got crashes. Now I cannot build and need to change our project files. Horrible.

mcosmin222 commented:
Yes, for the moment it seems to increase load on the CPU cores. I will try with the 4K hevc in the morning, that should see to my CPU taking off.

We can use the opaque pointer in av codec context to inject the config class inside that method.

The most annoying part about VS is that it cannot be rolled back easily.

mcosmin222 commented:
I can give you my build if you want.

mcosmin222 commented:
So, I manually forced a number of threads in get format method. This is with 12, which should occupy my entire CPU

![image](https://user-images.githubusercontent.com/9284164/92438324-435e9980-f1b1-11ea-9873-8b8b1feaa8dd.png)

In this image, it is with 4 threads

![image](https://user-images.githubusercontent.com/9284164/92438444-7dc83680-f1b1-11ea-8b72-b1bf7b618aa9.png)


I'd say it works, average utilization is lower with 4 threads (25%) compared to 12 (60%).
I will push the changes so you can test them as well.

lukasf commented:
Huh for me it does not really work. When I hard code thread_count to 8 in CreateVideoStream() already, and play a 4K HDR file, it plays almost smooth with nearly 100% CPU utilization. When I leave it as is, setting to 8 only later in get_frame, then I only get 25-30% utilization and very poor performance. So much like in HW acceleration, it seems that thread_count needs to be set correctly before opening the codec.

mcosmin222 commented:
But can we actually know for sure how many threads are used? Maybe this whole task manager CPU usage is not very accurate.

lukasf commented:
Now I have good news: I debugged into FFmpeg again, this time with FFmpeg 4.3.1, and now d3d11va_frames_init is only called once, even if thread_count is 8! So it seems that they fixed FFmpeg, to correctly ignore thread_count for HW decoding. I will undo the threading change and verify once again, but it looks like this will work now.



> 
> 
> But can we actually know for sure how many threads are used? Maybe this whole task manager CPU usage is not very accurate.

Playback was almost smooth when setting before open codec, and it was horribly laggy and slow when setting in get_frame. So I am very confident that multi threading was not in use. But anyways, it seems to work now.

mcosmin222 commented:
I am updating visual studio now for building with the other branch..

Nice find.

lukasf commented:
Oh, actually I wanted to ask you to build with an older VS version, to see if it is successful :)

mcosmin222 commented:
Too late :(

lukasf commented:
No worries. I think I can test this tomorrow on a different PC. I just have to see if all prerequisites are there.

lukasf commented:
Oh man. We get a crash on opening VC1 codec files. Looks like the VC1 decoder, unlike all others (and unlike written in the docs), will require the device pointers to be already set during codec open. It internally calls ff_get_format during its init step, which will already try to create the frames context from device context.

We'll have to see if we can somehow work around this. Maybe set default hw context, then clear it and later re-create. Not nice.

mcosmin222 commented:
Hmm maybe that one requires we provide a frames context ourselves.

mcosmin222 commented:
The video here: http://samples.mplayerhq.hu/V-codecs/WVC1/

Test_1440x576_WVC1_6Mbps.wmv    

reports weird timestamps. Seeking to any position seems to actually start playback from beginning, while timestamps don't reflect that.

lukasf commented:
We did not update frame pts. I fixed this and also made sure that all other frame properties are read and used (e.g. interlaced and other flags).

mcosmin222 commented:
Looks good now. Nice work.

I'm still very shocked that ffmpeg allows us to simply change that device context like that.

lukasf commented:
During codec open of VC1 and WMV3, FFmpeg executes ff_get_format() two times. Each time, any existing hw accel is destroyed and then a new a frames context and hw accel is created from the current device context. And on decode of the first frame, ff_get_format is executed for the third time, again destroying existing hw accel and creating it again. That seems pretty inefficient, three times creation of frames context and allocation and destroying of textures. But there is nothing we can do about that. The only good thing is that we can safely exchange the device context after codec open, because a new frames context will be created anyways during decode of the first frame.

It is a bit of a hack, but I don't see any good alternative. The only other approach would be to use the older approach where we copy the texture from default device pointer to mss device pointer. But I think that one has performance impact on higher resolutions, which I'd like to avoid.

mcosmin222 commented:
This looks like a bug in ffmpeg. We should be careful with this too make sure a fix won't break us. It seems this is quite a new implementation even for ffmpeg.

I always assumed get_format is called until a proper format is returned. The fact that wmv and vc1 do different sounds like a bug.

lukasf commented:
I think it's more like a pretty old implementation. I also wouldn't call it buggy since it actually works. It's just not super efficient. They re-use lots of old code from MSMPEG4 and H263, which makes the code pretty hard to understand. One of the unneccessary get_format calls comes from init calls for that older code.

ff_get_format does a lot more than just getting the format. It is also called on each flush(), also then it will destroy and re-init hw acceleration, to make sure it starts over with a fresh state.

Even if they change the behavior, to make it like H264 and others, our implementation would not break. In that case, we'd first create and assign a default device context, then destroy and replace it with the actuall mss context, all before it is used. We would just have one unneccessary creation of device context, which we could remove at some point.

mcosmin222 commented:
Ah that's good then. We give it a few more days to test and then we can merge I think...

mcosmin222 commented:
I also tried this with a media playback list, seems to work correctly.

mcosmin222 commented:
I'd say this is looking pretty well. Haven't found anything weird of late.

lukasf commented:
Agreed, merging this to master branch. I will still wait a while before creating an official NuGet packet.

villagra commented:
@lukasf would be possible to have -pre nuget packages with the hardware acceleration enabled to test it here with a bunch of files?

lukasf commented:
Hi @villagra 

sure, it is available already.

For the FFmpegInteropX library, you can add our [AppVeyor nuget feed](https://ci.appveyor.com/nuget/ffmpeginteropx) as nuget package source in VS. There you can easily get every automatic build from every branch (including master branch of course). Alternatively, you can download nuget packages from the artifacts section of a build (e.g. [here](https://ci.appveyor.com/project/lukasf/ffmpeginteropx/build/artifacts) for latest master), and manually install them using the VS package manager console.

For FFmpegInteropX.FFmpegUWP, there is 4.3.100-pre on the official nuget.org, which contains HW acceleration bits. It is mandatory if you want to try HW acceleration.

villagra commented:
Awesome, thanks!

lukasf commented:
I should maybe add that pull requests are not named after their feature branch, but after the last master branch into which they are merged for building plus a random suffix, e.g. *0.9.0-master99-[randomString]*. For a "normal" master branch build, take the latest that does not have that random suffix, e.g. *0.9.0-master99*.

Maybe I should add some information on this stuff on the main page.

villagra commented:
Thanks a lot!
So far It seems to be working great with all the movies I've tried... I'm gonna run a beta program with my app to get some feedback and I'll let you know if we find any issues.

lukasf commented:
Cool, any feedback would be welcome! It is quite a major change, so the more testing we get, the better it will be. I will still wait a while before making an official nuget package.

mcosmin222 commented:
I've integrated this into my app and it seems to work well so far.
